src/passes/TypeGeneralizing.cpp

```diff
@@ -0,0 +1,475 @@
+  }
+
+  void clearStack(Element& elem) const noexcept {
+    stack(elem) = stackLattice().getBottom();
+  }
+
+  const std::vector<Type>& getLocals(Element& elem) const noexcept {
+    return *locals(elem);
+  }
+
+  const std::vector<Type>& getLocals() const noexcept {
+    return *locals(getBottom());
+  }
+
+  Type getLocal(Element& elem, Index i) const noexcept {
+    return getLocals(elem)[i];
+  }
+
+  bool updateLocal(Element& elem, Index i, Type type) const noexcept {
+    return localsLattice().join(
+      locals(elem),
+      Vector<TypeRequirement>::SingletonElement(i, std::move(type)));
+  }
+
+private:
+  static LocalTypeRequirements initLocals(Function* func) noexcept {
+    return Shared{Vector{Inverted{ValType{}}, func->getNumLocals()}};
+  }
+
+  static ValueStackTypeRequirements initStack() noexcept {
+    return Stack{Inverted{ValType{}}};
+  }
+
+  const LocalTypeRequirements& localsLattice() const noexcept {
+    return std::get<LocalsIndex>(lattices);
+  }
+
+  const ValueStackTypeRequirements& stackLattice() const noexcept {
+    return std::get<StackIndex>(lattices);
+  }
+
+  typename LocalTypeRequirements::Element&
+  locals(Element& elem) const noexcept {
+    return std::get<LocalsIndex>(elem);
+  }
+
+  const typename LocalTypeRequirements::Element&
+  locals(const Element& elem) const noexcept {
+    return std::get<LocalsIndex>(elem);
+  }
+
+  typename ValueStackTypeRequirements::Element&
+  stack(Element& elem) const noexcept {
+    return std::get<StackIndex>(elem);
+  }
+
+  const typename ValueStackTypeRequirements::Element&
+  stack(const Element& elem) const noexcept {
+    return std::get<StackIndex>(elem);
+  }
+};
+
+struct TransferFn : OverriddenVisitor<TransferFn> {
+  Module& wasm;
+  Function* func;
+  State lattice;
+  typename State::Element* state = nullptr;
+
+  // For each local, the set of blocks we may need to re-analyze when we update
+  // the constraint on the local.
+  std::vector<std::vector<const BasicBlock*>> localDependents;
+
+  // The set of basic blocks that may depend on the result of the current
+  // transfer.
+  std::unordered_set<const BasicBlock*> currDependents;
+
+  TransferFn(Module& wasm, Function* func, CFG& cfg)
+    : wasm(wasm), func(func), lattice(func),
+      localDependents(func->getNumLocals()) {
+    // Initialize `localDependents`. Any block containing a `local.set l` may
+    // need to be re-analyzed whenever the constraint on `l` is updated.
+    auto numLocals = func->getNumLocals();
+    std::vector<std::unordered_set<const BasicBlock*>> dependentSets(numLocals);
+    for (const auto& bb : cfg) {
+      for (const auto* inst : bb) {
+        if (auto set = inst->dynCast<LocalSet>()) {
+          dependentSets[set->index].insert(&bb);
+        }
+      }
+    }
+    for (size_t i = 0, n = dependentSets.size(); i < n; ++i) {
+      localDependents[i] = std::vector<const BasicBlock*>(
+        dependentSets[i].begin(), dependentSets[i].end());
+    }
```

Comment:
Hmm, I was expecting this kind of locals dependency to be done in the framework. Was that wrong of me to hope for?

That is, this is likely a common pattern so I'm surprised to see it in this pass.

Reply:
Right now passes have to explicitly provide all the dependent blocks, including not only dependencies due to updated locals (and globals in the future), but also the basic predecessor / successor dependencies.

We have the `VisitorTransferFunction` that knows to add the predecessors or successors as dependencies for forward or backward analyses, but it doesn't know about locals or any other sources of dependencies.

Another place where you would hope the framework could be more helpful is with the transfer function for individual instructions. For example, `visitDrop` will simply do a `pop()` in the majority of backwards analysis. However, so far in this analysis, we only push reference types onto the stack, so a drop of an i32 should not do the `pop()`. I could imagine having a utility that applied a custom predicate and was able to do most of the pops automatically, but it wouldn't know what to push without user intervention.

I think the correct play is to get experience implementing a bunch of analyses, then figure out how to factor out layered utilities for the common patterns.